app/controllers/api/v1/messages_controller.rb

```diff
@@ -9,14 +9,17 @@ def create
       end
 
       def index
-        conversation = Conversation.find(params[:conversation_id])
-        raise InvalidConversationError unless conversation.user1_id == @current_user.id || conversation.user2_id == @current_user.id
-
-        render json: { messages: conversation.messages.page(params[:page]) }, status: :ok
+        @messages = IndexMessageService.new.index(params[:conversation_id], current_user, params[:page])
       rescue InvalidConversationError => e
         render json: { error: e.message }, status: :unprocessable_entity
       end
 
+      def unread_messages
+        conversations = Conversation.where(user1: current_user).or(Conversation.where(user2: current_user))
+        unread_messages = conversations.map { |conversation| { id: conversation.id, count: conversation.messages.where(user: conversation.opposite_user(current_user)).where(read: false).count } }
```

Comment:
maybe you can move `conversation.messages.where(user: conversation.opposite_user(current_user)).where(read: false).count` to a method on the conversation model and use it.
Also it's worth using jbuilder for this view